### DISCUSSION TITLE
Support for TM1814 with Arduino Nano

### DISCUSSION THREAD

**Discussion by Thagu** [Makuna/NeoPixelBus#603] [Q&A]
Hi Makuta

I've read the thread #325 about supporting TM1814 with ESP. Since I have an Arduino Nano I don't get it running: the [stripe ](https://de.aliexpress.com/item/1005004570700470.html?spm=a2g0o.order_list.0.0.13cc5c5f5du9Gh&gatewayAdapt=glo2deu)with the RGBW LEDs does something, but definitely not what I'm telling him. Most probably I need to build a circuit to invert the power/signal or so... But there my knowledge is not good enough to do so. Or is the current NeoPixelBus capable of running a TM1814 with an Arduino Nano without further electronics and I'm just to dumb?

I'd appreciate some support about the proper setup to get a TM1814 LED Stripe running with an Arduino Nano, even if this is a software forum. But I couldn't find anything good in the internet so far.
Thank you
Guido

**Comment by Makuna** [comment]
(moved to discussions as it wasn't an issue)
Use some Googlefu and look for "logic inverter".  There are plenty of hits.  Hardware is the way to solve this for the AVR platform (Nano).

**Reply by Thagu** [reply]
Thank you Makuna. I'll have a research about that. I see something similar [here ](https://create.arduino.cc/projecthub/mariogianota/control-a-digital-logic-not-chip-with-arduino-uno-d709e8)I guess.
For the further understanding: if a NOT inverter just puts out High if input is low (and vice versa), we can't we configure directly the Nano's port to do so? If I'm right there are pulldown and pullup options for this kind of stuff? Well, I'll miss something, I guess

**Reply by Makuna** [reply]
The code for AVR is assembly code, it would have to have a rewritten version for invert, this code is not mine and I reluctant to modify it as it would be harder to merge in changes from the source if it ever changed.
The AVR doesn't support a simple pin invert config like other platforms either.

**Reply by Makuna** [reply]
There are these NOT chips like you pointed at, but there are also simple circuits using a transistor.  Heck, often they can do the level conversion (3.3v to 5v) and invert the signal at the same time if you need it.

**Reply by Thagu** [reply]
Hi Makuna, I followed your words and established a NOT Gate via a transistor, following that instruction [here](https://www.youtube.com/watch?v=Q-sTB6aA6O4). Code on his blog side works, i.e. I can read the inverted signal with the arduino itself (on another pin). But with the neopixelbus i have no change in behaviour: basically no reaction. I used a 327-25 PNP transistor. Any ideas why this? I use D6 output as defined in the code. 
I use the inverted method:
NeoPixelBus<NeoWrgbTm1814Feature, NeoTm1814InvertedMethod > strip(PixelCount, PixelPin);

the non-inverted leads to an error that the method is not defined.

Am I on the wrong path?
Thank you in advance
Guido

**Reply by Thagu** [reply]
Possibly one issue could be that the stripes runs at 24v (on +/-), but arduino nano delivers 5V for the signal wire? (although this was never the case with other rgb stripes). (btw: I have connected both ground cable, i.e. from 24v power supply and arduino. )
I now made a test after having found a wiring issue with the transistor. but still no good result. If i connect now the led stripe directly to output 6 (i.e. without the not gate), then the first set of pixels (address 1) starts to flicker with the trial code. but all other leds still go through their default loop (by changing the color every 1s or so). I've adressed 300 leds in my script...

**Reply by Makuna** [reply]
The strip voltage versus logic voltage are two different things.  Connecting the GNDs is required.  Check the spec sheet, but I don't believe the logic level has to match the driving voltage at all; but it may list the range it has to be in.


You didn't forget this line after you call strip.Begin?
```
// 16.5 ma for each RGB channel, and 38 ma for the white channel
strip.SetPixelSettings(NeoTm1814Settings(165,165,165,380)); // RGBW currents, in tenth milliamperes
```
Of course, use values that match your LEDs for current draw so you don't over drive them.  
Not having this line defaults the current to the lowest the chips support, which depending on your actual LEDs may not show anything.

**Reply by Thagu** [reply]
I've pretty much copied your code you posted somewhere in the forum, but changed the pixel colouring to all pixels and not just the first few:

```
#include <NeoPixelBus.h>

const uint16_t PixelCount = 300; // this example assumes 4 pixels, making it smaller will cause a failure
const uint8_t PixelPin = 10; // make sure to set this to the correct pin, ignored for Esp8266

NeoPixelBus<NeoWrgbTm1814Feature, NeoTm1814InvertedMethod > strip(PixelCount, PixelPin);
//NeoPixelBus<NeoWrgbTm1814Feature, NeoTm1814Method > strip(PixelCount, PixelPin);

// colours normally White, Red, Green, and then Blue

void setup() {
  strip.Begin();
//  strip.SetPixelSettings(NeoTm1814Settings(165,165,165,380));  //set milliamps to be used
  strip.SetPixelSettings(NeoTm1814Settings(80,80,80,80));  //set milliamps to be used
  strip.Show();
  delay(50);
}

void loop() {
  for (int i = 0; i<PixelCount; i++){
    strip.SetPixelColor(i, RgbwColor(20,40,40,40)); // red, auto converts to RgbwColor that these strips need
  }
  strip.Show();
  delay(100);
  for (int i = 0; i<PixelCount; i++){
    strip.SetPixelColor(i, RgbwColor(0,0,255,0)); // red, auto converts to RgbwColor that these strips need
  }
  strip.Show();
  delay(100);
}
```


I set the current to 8mA, which sould be below the limit of 10 or 15 mA. I went above and below - no change. I also tried with your original script - with a few leds and dedicated colours. 
What I realised is, when I plug to the non-inverted signal (i.e. directly to pin 10) the strip stops changing the colour and remains at the current one. With full power.

I've tried two different stripes, same behaviour. Thus, shouldn't be a defect of the chips.

**Reply by Thagu** [reply]
I double check: the chip expects 5V for DI, but possibly the stripe is doing somehting else. I also wrote a request to the seller for some support.

**Reply by Thagu** [reply]
Hi Community and Makuna, 
I sitll haven't got these stripes working. Any idea which next trials are the most promising? Different inverter? 

Thank you for any kind of input.

Regards, Guido

**Reply by Makuna** [reply]
You will need the inverting circuit.  Make sure it fast enough, which means it can switch off and on very quickly.  You need signal quality transistors/parts.

You haven't sent a link the actual strip.  Are they really RGBW?  Do the strips have two data inputs?  If so, connect both to the signal pin.  

Keep things simple until you get it working.  Change back the sample to only modifying the first 4 pixels.  Change the strip length to 4 (doesn't have to match the actual strip, colors after the first 4 maybe a little random).  

What is the memory left after the compilation of the sketch (Arduino lists this when it's done compiling).

**Reply by Thagu** [reply]
Hi Makuna

thank you for the input. Thought I posted the link to the strip, though this wasn't the case. Here it is: [Strip](https://de.aliexpress.com/item/1005004570700470.html?spm=a2g0o.order_list.order_list_main.11.38e65c5fk2oluN&gatewayAdapt=glo2deu)

It is RGBW, that's why I took that kind of stip with TM1814. (Was the only one I found with 24 Volt...) There is only 1 data cable available.

Since I was using the 327-25 Transistor, I'll look for a faster one. Does anyone have a dedicated proposal for a good transistor? Since I'll need to order it online the packaging&transport is pretty cost intensive. So I don't want to go for trial & error :-)

**Reply by Thagu** [reply]
I'll compile the code later on, will need to switch to my private computer. Btw, I always checked also the simple script with only 4 LEDs. Since the leds are so bright is difficult to look at a signle LED and to detect what it is doing - that's why I extended it to 300 LEDs.

**Reply by Thagu** [reply]
ok, just compiled: sketch uses 2420 bytes (7%) ofr storage space and 34 bytes (1%) of dynamic memory...

**Reply by Makuna** [reply]
Do you happen to have a Esp8266, Esp32, NRF52x (Nano 33 BLE)?  These support internal signal inverting, and you could test that its not the inverting circuit.

**Reply by Thagu** [reply]
nope, neither have I any experiences with it. I think I rather stay with the Nano. Building up the infrastructure would cost me a lot of time wihtout the knowledge.

**Reply by Makuna** [reply]
My next suggestion is either use an oscilloscope and/or logic analyzer to look at the signal before and after the inverter.  Some problems are caused by poor signal form, spikes on edges of pulses and rounding of the edges causing them to be interpreted incorrectly; these caused by too long of wires, not good filtering (capacitors), signal interference on wiring (twist signal and ground from nano to strip; add resistor in line).

Oh, and connecting gnd between nano and 24v power.

**Reply by Thagu** [reply]
I got gnd connected. I have no Oscilloscope (and even would have issues in handling it) nor a logic analyzer (never heard about that one. I'd rather try to order some good capacitor and try with them (even if it takes a while to get them). Any proposal? Or shall I go for the 74HC14 that was mentioned in the other thread?

**Reply by Thagu** [reply]
Goood morning - in the mean time I got introduced to the world of ESP32 boards... ordered, received and installed. And the library works with the basic code :-) Means, the strip is running. So I give up the path with the arduino but rather take the ESP32 board. I need to change my current code a bit - since my version is not working yet with the strip. But I guess it's a promising path. I'll get back to you, once it's solved or I have further questions. Thank you for the help until now.
